### PR TITLE
Copter: add pre-arm and mode init checks of ekf's altitude estimate

### PR DESCRIPTION
**Pull request by rmackay9**
This PR resolves issue https://github.com/ArduPilot/ardupilot/issues/14757 by adding a mandatory pre-arm check of the AHRS/EKF altitude estimate.  A check is also added to the set_mode call to catch problems when the user attempts to switch modes.

The issue is more serious than initially expected if the user sets EKx_ALT_SOURCE = 2 (GPS) because the user could more easily arm and takeoff in AltHold before the EKF gets a good height estimate.  This can lead to the vehicle continuously climbing or descending at it's maximum rate (i.e. 2.5m/s or 1.5m/s) which at best leads to a hard landing.

This has been tested in SITL including reproducing the existing behaviour in master (see below)
copter-alt-source2-althold

Below is a screen shot of some SITL testing in which arming and switching modes while flying was prevented.
image

The only small issue we may have is that both the, "Need Position Estimate" and "Need Alt Estimate" messages may be displayed before there is a GPS lock and EKx_ALT_SOURCE = 2.

### CODE REVIEW

File: ArduCopter/mode.cpp
Comment (khancyr) [inline ArduCopter/mode.cpp:259]: require alt estimate ? require EKF alt ? 
"Require alt" is quite disturbing

**Review by khancyr — APPROVED** [review]
Beside my comment on the "requite alt" it looks good.

I will just try to replicate on SITL and check the fix on SITL ... as it is a safety function, I think I would be good to make a test for this

File: ArduCopter/mode.cpp
Comment (rmackay9) [reply ArduCopter/mode.cpp]: great, thanks, I've updated this to "needs alt estimate" which is the same wording as the pre-arm failure.

**Comment by rmackay9** [comment]
@lthall, do you want to check this before we merge?  No huge pressure of course I think it's safe.

**Comment by khancyr** [comment]
Ok I have replicate the issue and the fix ! 
```
param set EK2_ALT_SOURCE 2
param set SIM_GPS_DISABLE 1
arm throttle
mode alt_hold
rc 3 2000
```
Wait like 5s that you takeoff and 
`rc 3 1500`
the drone keep climbing... same for `rc 3 1200` it will start slowdown and then keep going down !

The fix is working as intend ! I will make a PR for a test tomorrow

**Comment by rmackay9** [comment]
I've rebased on master and fixed up a comment.  I think I'll merge this after it passes travis.  Thanks @khancyr for the review!

**Pull request merged**
